eng/common/cross/gen-buildsys.sh

```diff
@@ -0,0 +1,190 @@
+  echo "Use the Ninja generator instead of the Unix Makefiles generator."
+  echo "Pass additional arguments to CMake call."
+  exit 1
+fi
+
+tryrun_dir="$2"
+build_arch="$4"
+compiler="$5"
+cxxCompiler="$compiler++"
+majorVersion="$6"
+minorVersion="$7"
+
+if [ "$compiler" = "gcc" ]; then cxxCompiler="g++"; fi
+
+check_version_exists() {
+    desired_version=-1
+
+    # Set up the environment to be used for building with the desired compiler.
+    if command -v "$compiler-$1.$2" > /dev/null; then
+        desired_version="-$1.$2"
+    elif command -v "$compier$1$2" > /dev/null; then
+        desired_version="$1$2"
+    elif command -v "$compiler-$1$2" > /dev/null; then
+        desired_version="-$1$2"
+    fi
+
+    echo "$desired_version"
+}
+
+if [ -z "$CLR_CC" ]; then
+
+    # Set default versions
+    if [ -z "$majorVersion" ]; then
+        # note: gcc (all versions) and clang versions higher than 6 do not have minor version in file name, if it is zero.
+        if [ "$compiler" = "clang" ]; then versions=( 9 8 7 6.0 5.0 4.0 3.9 3.8 3.7 3.6 3.5 )
+        elif [ "$compiler" = "gcc" ]; then versions=( 9 8 7 6 5 4.9 ); fi
+
+        for version in "${versions[@]}"; do
+            parts=(${version//./ })
+            desired_version="$(check_version_exists "${parts[0]}" "${parts[1]}")"
+            if [ "$desired_version" != "-1" ]; then majorVersion="${parts[0]}"; break; fi
+        done
+
+        if [ -z "$majorVersion" ]; then
+            if command -v "$compiler" > /dev/null; then
+                if [ "$(uname)" != "Darwin" ]; then
+                    echo "WARN: Specific version of $compiler not found, falling back to use the one in PATH."
+                fi
+                export CC="$(command -v "$compiler")"
+                export CXX="$(command -v "$cxxCompiler")"
+            else
+                echo "ERROR: No usable version of $compiler found."
+                exit 1
+            fi
+        else
+            if [ "$compiler" = "clang" ] && [ "$majorVersion" -lt 5 ]; then
+                if [ "$build_arch" = "arm" ] || [ "$build_arch" = "armel" ]; then
+                    if command -v "$compiler" > /dev/null; then
+                        echo "WARN: Found clang version $majorVersion which is not supported on arm/armel architectures, falling back to use clang from PATH."
+                        export CC="$(command -v "$compiler")"
+                        export CXX="$(command -v "$cxxCompiler")"
+                    else
+                        echo "ERROR: Found clang version $majorVersion which is not supported on arm/armel architectures, and there is no clang in PATH."
+                        exit 1
+                    fi
+                fi
+            fi
+        fi
+    else
+        desired_version="$(check_version_exists "$majorVersion" "$minorVersion")"
+        if [ "$desired_version" = "-1" ]; then
+            if command -v "$compiler" > /dev/null; then
```

Comment:
A nit - my preference in the case the requested version of the compiler wasn't installed is to error instead of trying to fall back.

Reply:
Sure, removed the fallback.

Reply:
I had to make a change in pipeline YAML for AppleClang: https://github.com/dotnet/runtime/pull/1072/commits/02ed5186f106348c2009165c44e44a2a4d10c8d1#diff-45d0c31b08d09fe299a0b89aba6ab31d, as the master branch is relying on this fallback:

https://github.com/dotnet/runtime/blob/ca0490944385bbb3dd875df0aa7e51a3d1d9c9be/src/coreclr/src/pal/tools/find-clang.sh#L24-L26